includes/sanitizers/class-amp-style-sanitizer.php

```diff
@@ -349,230 +431,741 @@ private function process_link_element( DOMElement $element ) {
+		$css_list->append( $important_ruleset ); // @todo It would be preferable if the important ruleset were inserted adjacent to the original rule.
+
+		return $validation_errors;
 	}
 
 	/**
-	 * Filter individual CSS name/value pairs.
+	 * Collect and store all CSS style attributes.
 	 *
-	 *   - Remove overflow if value is `auto` or `scroll`
-	 *   - Change `width` to `max-width`
-	 *   - Remove !important
+	 * Collects the CSS styles from within the HTML contained in this instance's DOMDocument.
+	 *
+	 * @see Retrieve array of styles using $this->get_styles() after calling this method.
 	 *
 	 * @since 0.4
+	 * @since 0.7 Modified to use element passed by XPath query.
+	 *
+	 * @note Uses recursion to traverse down the tree of DOMDocument nodes.
 	 *
-	 * @param string $property Property.
-	 * @param string $value    Value.
-	 * @return array
+	 * @param DOMElement $element Node.
 	 */
-	private function filter_style( $property, $value ) {
-		/*
-		 * Remove overflow if value is `auto` or `scroll`; not allowed in AMP
-		 *
-		 * @todo This removal needs to be reported.
-		 * @see https://www.ampproject.org/docs/reference/spec.html#properties
-		 */
-		if ( preg_match( '#^overflow(-[xy])?$#i', $property ) && preg_match( '#^(auto|scroll)$#i', $value ) ) {
-			return array( false, false );
+	private function collect_inline_styles( $element ) {
+		$style_attribute = $element->getAttributeNode( 'style' );
+		if ( ! $style_attribute || ! trim( $style_attribute->nodeValue ) ) {
+			return;
 		}
 
-		if ( 'width' === $property ) {
-			$property = 'max-width';
+		$class  = 'amp-wp-' . substr( md5( $style_attribute->nodeValue ), 0, 7 );
+		$root   = ':root' . str_repeat( ':not(#_)', 5 ); // @todo The correctness of using "5" should be validated.
+		$rule   = sprintf( '%s .%s { %s }', $root, $class, $style_attribute->nodeValue );
+		$hash   = md5( $rule );
+		$rule   = $this->process_stylesheet( $rule, $style_attribute, array(
+			'convert_width_to_max_width'  => true,
+			'allowed_at_rules'            => array(),
+			'property_whitelist'          => $this->style_custom_cdata_spec['css_spec']['allowed_declarations'],
+			'class_selector_tree_shaking' => false,
+		) );
+		$length = strlen( $rule );
+
+		if ( 0 === $length ) {
+			$element->removeAttribute( 'style' );
+			return;
 		}
 
-		/*
-		 * Remove `!important`; not allowed in AMP
-		 *
-		 * @todo This removal needs to be reported.
-		 */
-		if ( false !== strpos( $value, 'important' ) ) {
-			$value = preg_replace( '/\s*\!\s*important$/', '', $value );
+		if ( $this->current_custom_size + $length > $this->style_custom_cdata_spec['max_bytes'] ) {
+			$this->remove_invalid_attribute( $element, $style_attribute, array(
+				/* translators: %d is the number of bytes over the limit */
+				'message' => sprintf( __( 'Too much CSS enqueued (by %d bytes).', 'amp' ), ( $this->current_custom_size + $length ) - $this->style_custom_cdata_spec['max_bytes'] ),
+			) );
+			return;
```

Comment:
This needs to be eliminated. At the very least it should be done after all stylesheets have been gathered and filtered (tree shook). But even then it can result in broken styles if the styles are removed.